[WPE][GTK][WebDriver][BiDi] Implement the browser.close command

yury-s opened the pull request:
#### b455e41fb59d4599c6af7c5ec0326c672d358706
<pre>
[WPE][GTK][WebDriver][BiDi] Implement the browser.close command
<a href="https://bugs.webkit.org/show_bug.cgi?id=288103">https://bugs.webkit.org/show_bug.cgi?id=288103</a>

Reviewed by NOBODY (OOPS!).

Added basic implementation of browser.close() command that closes
the application process.

* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/APIAutomationClient.h:
(API::AutomationClient::closeBrowser):
* Source/WebKit/UIProcess/API/glib/WebKitBrowserAutomation.cpp: Added.
A new browser wide automation interface. It can be used with WebDriver Bidi
where Automation spans across WebKitWebContext&apos;s rather than being bound
to a single context as glib/WebKitAutomationSession today. The same interface
can be extended later with signals for creating pages in particular user
contexts and creating contexts for automation.

(WEBKIT_DEFINE_FINAL_TYPE):
(webkit_browser_automation_class_init):
(createDefaultBrowserAutomation):
(webkitBrowserAutomationMaybeSetClient):
(webkit_browser_automation_get_default):
* Source/WebKit/UIProcess/API/glib/WebKitBrowserAutomation.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitBrowserAutomationPrivate.h: Added.
* Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
AutomationClient is currently owned by WebProcessPool. This will need to be
changed to support WebDriver Bidi browser module that operates on multiple
user contexts.

* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp:
(WebKit::BidiBrowserAgent::close):
The agent calls into the automation client, eventually the client will need
to be global per browser rather than bound to a process pool.

* Source/WebKit/UIProcess/WebProcessPool.h:
* Tools/MiniBrowser/gtk/main.c:
(quitApplicationForAutomation):
(main):
* Tools/MiniBrowser/wpe/main.cpp:
(quitApplicationForAutomation):
(main):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b455e41fb59d4599c6af7c5ec0326c672d358706

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/104201 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/23905 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/14231 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/109399 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/54869 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/106241 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/24330 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/32449 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/79142 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/107207 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/18906 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/94012 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/59469 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/18713 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/12048 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/54229 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/88415 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/12107 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/111783 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/31357 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/23119 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/88165 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/31721 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/90199 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/87826 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/32716 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/10470 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/25821 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/31286 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/36599 "Failed to build and analyze WebKit") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/31080 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/34416 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/32640 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

burg reviewed (approved):
Looks good to me, though it would be good to have a glib reviewer take a look as well.

I'm still unconvinced we need to support multiple process pools to get multiple network sessions. I'll try to confirm when I implement the Mac and iOS parts of user context commands.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitBrowserAutomation.cpp:
```diff
@@ -0,0 +1,128 @@
+        0,
+#else
+        G_STRUCT_OFFSET(WebKitBrowserAutomationClass, quit_application),
+#endif
```
Let's avoid exposing the class struct, so this can be just:

```suggestion
        0,
```

yury-s replied:
Done.

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitBrowserAutomation.h.in:
```diff
@@ -0,0 +1,58 @@
+};
+#else
+WEBKIT_DECLARE_FINAL_TYPE (WebKitBrowserAutomation, webkit_browser_automation, WEBKIT, BROWSER_AUTOMATION, GObject)
+#endif
```
The boilerplate is needed in other classes for backwards-compatibility, but this class is completely new, so we don't need it:

```suggestion
WEBKIT_DECLARE_FINAL_TYPE (WebKitBrowserAutomation, webkit_browser_automation, WEBKIT, BROWSER_AUTOMATION, GObject)
```

yury-s replied:
Done.

mcatanzaro reviewed (changes requested):
Question: how does WebKitBrowserAutomation differ from WebKitAutomationSession? They seem very similar. I see that `WebKit.AutomationSession::will-close` is not _quite_ the same as `WebKitBrowserAutomation::quit-application`, but I think the documentation should explain the differences between both the classes and the signals to avoid confusion; each should probably explain why it is different than the other. Is the idea that an application might have many automation sessions, but only one "browser automation"? Is it not possible to combine the functionality into WebKitAutomationSession alone? Are you planning to add more stuff that would belong in WebKitBrowserAutomation but not WebKitAutomationSession?

The other thing I don't like here is that `WebKitBrowserAutomation` is a singleton. Is there a strong reason this can't be tied to a `WebKitWebContext` instead? `webkit_web_context_get_browser_automation()` would be nicer; that should also resolve your FIXME comment complaining that the close operation is tied to a particular process pool, but shouldn't be. Think of `WebKitWebContext` as the library context object: it should allow a single application to run entirely separate instances of WebKit. (Or mostly entirely separate; yes, it's true they'll share a network process by default.) Global data causes too many difficult problems (e.g. https://bugs.webkit.org/show_bug.cgi?id=284807) and we should avoid it unless it's truly needed.

@carlosgcampos could you review this please?

lauromoura commented on Source/WebKit/UIProcess/API/glib/WebKitBrowserAutomation.cpp:
```diff
@@ -0,0 +1,128 @@
+    GObjectClass* gObjectClass = G_OBJECT_CLASS(automationClass);
+
+    /**
+     * WebKitBrowserAutomation::quit-application:
```
What about the name `close-browser`, to match the underlying BiDi command?

yury-s replied:
Renamed.

lauromoura commented on Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp:
```diff
@@ -80,6 +81,10 @@ Inspector::CommandResult<void> BidiBrowserAgent::close()
 
+    // FIXME: this should not depend on a particular process pool.
+    if (auto* client = session->protectedProcessPool()->automationClient())
+        client->closeBrowser();
```
We could return the `"unable to close browser"` error if client is not set, with an error message like "Could not find automation client" or similar. From the spec:

> "unable to close browser": Tried to close the browser, but failed to do so.

yury-s replied:
Done.

lauromoura reviewed (commented):
Looks good in general. Two minor comments.

yury-s committed:
[WPE][GTK][WebDriver][BiDi] Implement the browser.close command
https://bugs.webkit.org/show_bug.cgi?id=288103

Reviewed by NOBODY (OOPS!).

Added basic implementation of browser.close() command that closes
the application process.

* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/APIAutomationClient.h:
(API::AutomationClient::closeBrowser):
* Source/WebKit/UIProcess/API/glib/WebKitBrowserAutomation.cpp: Added.
A new browser wide automation interface. It can be used with WebDriver Bidi
where Automation spans across WebKitWebContext's rather than being bound
to a single context as glib/WebKitAutomationSession today. The same interface
can be extended later with signals for creating pages in particular user
contexts and creating contexts for automation.

(WEBKIT_DEFINE_FINAL_TYPE):
(webkit_browser_automation_class_init):
(createDefaultBrowserAutomation):
(webkitBrowserAutomationMaybeSetClient):
(webkit_browser_automation_get_default):
* Source/WebKit/UIProcess/API/glib/WebKitBrowserAutomation.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitBrowserAutomationPrivate.h: Added.
* Source/WebKit/UIProcess/API/glib/WebKitWebContext.cpp:
AutomationClient is currently owned by WebProcessPool. This will need to be
changed to support WebDriver Bidi browser module that operates on multiple
user contexts.

* Source/WebKit/UIProcess/API/glib/webkit.h.in:
* Source/WebKit/UIProcess/Automation/BidiBrowserAgent.cpp:
(WebKit::BidiBrowserAgent::close):
The agent calls into the automation client, eventually the client will need
to be global per browser rather than bound to a process pool.

* Source/WebKit/UIProcess/WebProcessPool.h:
* Tools/MiniBrowser/gtk/main.c:
(quitApplicationForAutomation):
(main):
* Tools/MiniBrowser/wpe/main.cpp:
(quitApplicationForAutomation):
(main):
